### ISSUE TITLE
missing `model.wv.index_to_key` and mismatch node number with index vector

### ISSUE THREAD

**Issue by yasirroni** [eliorc/node2vec#52]
Using `model.wv.vectors` we can extract list of vectors. Unfortunately, the index is unsorted and doesn't preserve node indexing.

To see the paired `node` (label) with its corresponding `vector`, we can use `model.wv.save_word2vec_format('word2vec_format.json')`. I think this is a bug.

**Comment by yasirroni** [comment]
For reproduction:

```python

import networkx as nx

## edges_list
edges_list = [
    (1, 2, 1), 
    (2, 3, 1), 
    (3, 4, 2), 
    (4, 5, 1), 
    (5, 6, 1), 
    (1, 7, 1), 
    (7, 8, 1), 
    (8, 9, 4), 
    (9, 10, 1), 
    (8, 11, 1), 
    (11, 12, 2),
]

## initiate directed graph
dg = nx.DiGraph() # directed graph
dg.add_weighted_edges_from(edges_list)

import random
import numpy as np
from node2vec import Node2Vec

SEED = 42
random.seed(SEED)
np.random.seed(SEED)

# # Create a graph
# graph = nx.fast_gnp_random_graph(n=100, p=0.5)

params = {
    'graph' : dg,
    'weight_key': dg.edges.data('weight'),
    'p': 2,
    'q': 0.5,
    'dimensions' : 2,
    'walk_length' : dg.number_of_nodes() // 0.8, # 14
    'num_walks' : 10000,
    'workers' : 1,
}

# Precompute probabilities and generate walks
node2vec = Node2Vec(**params)

# Embed nodes
model = node2vec.fit(window=2, min_count=1)

from sklearn.preprocessing import MinMaxScaler
scaler = MinMaxScaler()
vectors = scaler.fit_transform(model.wv.vectors)
print(vectors)

model.wv.save_word2vec_format('word2vec_format.json')
```

**Comment by yasirroni** [comment]
Index name actually saved in `model.wv.index2word`.

This behavior needs to be documented or auto sort needs to be implemented.

**Comment by yasirroni** [comment]
I made this:

```python
def get_word2vec_dict(model, scaler=None):
    if scaler:
        vectors = scaler.fit_transform(model.wv.vectors)
        vectors = vectors.tolist()
    else:
        vectors = model.wv.vectors.tolist()
        
    word2vec_dict = {}
    for label, vector in zip(model.wv.index2word, vectors):
        word2vec_dict[label] = vector 
    return word2vec_dict
```
usage:

```python
from sklearn.preprocessing import MinMaxScaler
scaler = MinMaxScaler()
word2vec_dict = get_word2vec_dict(model, scaler=scaler)
word2vec_dict
```

Should I make a PR, or is it has been implemented?

**Comment by eliorc** [comment]
Hi @yasirroni 

What yo uget in return after calling `node2vec.fit` is actually an object from another library - so the `model` object is actually not under my control - it belong to the following repository 

https://github.com/RaRe-Technologies/gensim

**Comment by yasirroni** [comment]
> Hi @yasirroni
> 
> What yo uget in return after calling `node2vec.fit` is actually an object from another library - so the `model` object is actually not under my control - it belong to the following repository
> 
> https://github.com/RaRe-Technologies/gensim

Thank you for the reply.

I understand that. That's why I also lookup in Gensim for any information regarding model manipulation and extraction. But, node2vec is a special case of graph, not just a bunch of words.
_
Regarding word2vec_dict, I propose that it's added on node2vec manipulation tools. For example:

from node2vec import node2vec as nv


word2vec_dict = nv.get_word2vec_dict(gensim_model)

Other utils tools can go there too.

**Comment by eliorc** [comment]
I see how this can help, but since this is regarding the manipulation of `gensim`'s structure implementing this under node2vec doesn't feel right to me

But regarding your situation, I don't really understand your problem - why don't you just use `model.wv['<your required node>']` to access vectors? Why do you build a custom dictionary?

**Comment by yasirroni** [comment]
We might not need to modify gensim model structure. The output can bes set as separate object.
_
Regarding the `model.wv['<node>']`, I don't know about that. I think it solve my problems. But, since I need vector manipulation (and also don't want to change gensim model), i still need that function to make separate object to doo vector manipulation such as dimensions reduction using tsne or normalisation.

We can close this issue if you didn't interested to include some utils to do vector manipulation or plot wrapper in the library.

**Comment by yasirroni** [comment]
Note:

'model.wv.vectors' still need to be included in readme.md to warn other user that the indexing doesn't preserve node numbering.

**Comment by eliorc** [comment]
> `model.wv.vectors` still need to be included in `readme.md` to warn other user that the indexing doesn't preserve node numbering.

I think this should be apparent to the user by reading `gensim`'s docs

Thanks you for clarifying things and I hope collaborators with similar questions will arrive in this issue
